SEAB-6148: Log more diagnostic information

svonworl opened the pull request:
**Description**
This PR introduces a new `DiagnosticsHelper` class that the webservice application can wire into itself and log various information:

* Periodic "Global" info about ~~running processes,~~ the filesystem, the database, and memory pools, logged via a `Timer` daemon thread.
*  Request-level info about a request and associated thread and Hibernate `Session`, logged at the start and finish of each request, two log entries per request.

In this iteration, I dropped the thread dump from the periodic logging output.  At 25kB a pop, it didn't seem worth it.

Diagnostic logging is enabled using the `diagnosticsConfig` configuration file property.  Periodic and request logging can be enabled/disabled separately, and the logging period can be set.  By default, diagnostic logging is disabled.  To enable both periodic and request logging, add the following to your config file:

```
diagnosticsConfig:
  logRequests: true
  logPeriodic: true
  periodSeconds: 60
```

To review the output, pipe the webservice stdout into `tee /tmp/output` and view the resulting file.  Search for `DiagnosticsHelper` to find the new log entries.

~~A big concern is that, since we're logging data pulled out of the guts of our runtime environment, we'll accidentally log a secret, especially via the `ps` output, which includes the arguments to commands.  We purposefully do not log environment variables, which are a known vector for secrets.~~

~~To reduce the chance of a secret spilling, all of the information that is logged - literally the entirety of each log message - passes through censoring code that looks for runs of characters that appear to be hex/base64, does some analysis on the character distributions, and X's them out if they appear to be keys.  The crux here is that file paths resemble base64 strings because they tend to contain the same characters, so we need a way to let them through uncensored.  The assumption is that a key will be well-encoded and contain a random jumble of characters, whereas a path (or other structured non-key data) will contain longer runs of the same character classes, and/or contain text that resembles English.  The resulting censoring scheme works pretty well, see the comments on the `CensorHelper` class for more information.~~

Another concern is that these changes will introduce instability into the webservice.  Keep an eye out for that type of problem.  We're purposefully doing the periodic logging via a `Timer`/thread so in case it hangs or throws, the normal functions of the webservice aren't as likely to be impacted. 

We can easily add a filter that will exclude certain requests from diagnostic logging (for example, hits to the health check endpoint).

**Review Instructions**
View the Cloudwatch logs for qa, and confirm that diagnostic log entries are being created before and after each request, and periodically every so often.  Search for "DiagnosticsHelper" to find the entries.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-6148

**Security and Privacy**
Many security concerns, see description.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

denis-yuen commented:
> Basically, this PR introduces a new `DebugHelper` class which wires itself into the infrastructure and logs various information. "Global" info about threads, processes, the database, and memory is logged periodically via a `Timer`, and request-level info about a handling thread and Hibernate `Session` is logged at the end of each request.

Seems like a lot of data, not having looked at the code yet, wonder if it should be logged to a separate appender/file
https://www.dropwizard.io/en/stable/manual/configuration.html#logging
that could eventually end up in a log group so we can more easily choose what to view and/or have different expiry/deletion rules

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DebugHelper.java:
```diff
@@ -0,0 +1,337 @@
+
+            // Request started.
+            case START:
+                logStarted(request);
```
This will be very useful; right now the request is only logged when completed. If we die (rare, but it's happened) processing a request, it's hard to piece together what the request was.

coverbeck reviewed (commented):
Conceptually, looks good.

Do you have some sample output, just to get a sense of what it might look like and how large it would be?

> > Basically, this PR introduces a new `DebugHelper` class which wires itself into the infrastructure and logs various information. "Global" info about threads, processes, the database, and memory is logged periodically via a `Timer`, and request-level info about a handling thread and Hibernate `Session` is logged at the end of each request.
> 
> Seems like a lot of data, not having looked at the code yet, wonder if it should be logged to a separate appender/file https://www.dropwizard.io/en/stable/manual/configuration.html#logging that could eventually end up in a log group so we can more easily choose what to view and/or have different expiry/deletion rules

On the one hand, yes it could make the "main" log really noisy and might be worth separating. On the other hand, if it would make easier correlating things if we're researching an issue, e.g., seeing the thread dump right before/after a request is logged would be useful. I wonder if CloudWatch has a way to "merge" log groups when querying?

denis-yuen commented:
> I wonder if CloudWatch has a way to "merge" log groups when querying?

Looks like, yes, via the insights queries 
https://aws.amazon.com/about-aws/whats-new/2019/07/cloudwatch-logs-insights-adds-cross-log-group-querying/

svonworl commented:
> Conceptually, looks good.
> 
> Do you have some sample output, just to get a sense of what it might look like and how large it would be?
> 
> > > Basically, this PR introduces a new `DebugHelper` class which wires itself into the infrastructure and logs various information. "Global" info about threads, processes, the database, and memory is logged periodically via a `Timer`, and request-level info about a handling thread and Hibernate `Session` is logged at the end of each request.
> > 
> > 
> > Seems like a lot of data, not having looked at the code yet, wonder if it should be logged to a separate appender/file https://www.dropwizard.io/en/stable/manual/configuration.html#logging that could eventually end up in a log group so we can more easily choose what to view and/or have different expiry/deletion rules
> 
> On the one hand, yes it could make the "main" log really noisy and might be worth separating. On the other hand, if it would make easier correlating things if we're researching an issue, e.g., seeing the thread dump right before/after a request is logged would be useful. I wonder if CloudWatch has a way to "merge" log groups when querying?

Some thoughts:
* It's relatively difficult to send separate streams of information to different log groups from a container running in ECS.  The typical setup is to use the cloudwatch agent to send stdout/stderr to a group.
* We could possibly modify our logging to condense each multi-line log entry to a single expandable entry: https://stackoverflow.com/questions/67396522/cloudwatch-multiline-log-messages-from-containerized-app-runnning-on-ecs-ec2  This would greatly reduce the visual impact of the logging entries that appear in this PR, as well as collapse the giant stack traces that currently appear in our logs.
* It is most useful to display the debugging information interleaved temporally with our other log output.  The hard it is to do that, the less useful the debug information is.

Regarding size, there's two groups of information being logged.  The first, thread/request-level info, is pretty small:
```
INFO  [2024-02-28 20:50:59,095] io.dockstore.webservice.helpers.DebugHelper: debug.started by thread "dw-45 - GET /organizations/9/starredUsers" (45):
GET "organizations/9/starredUsers"
INFO  [2024-02-28 20:50:59,109] io.dockstore.webservice.helpers.DebugHelper: debug.session by thread "dw-45 - GET /organizations/9/starredUsers" (45):
SessionStatistics[entity count=1,collection count=4]
INFO  [2024-02-28 20:50:59,131] io.dockstore.webservice.helpers.DebugHelper: debug.finished by thread "dw-45 - GET /organizations/9/starredUsers" (45):
GET "organizations/9/starredUsers"
INFO  [2024-02-28 20:50:59,131] io.dockstore.webservice.helpers.DebugHelper: debug.thread by thread "dw-45 - GET /organizations/9/starredUsers" (45):
allocated: 0.60 MB
cpu-time: 0.012 sec
user-time: 0.011 sec
elapsed-time: 0.036 sec
```

The periodic "global" information is larger:
```
INFO  [2024-02-28 20:51:05,045] io.dockstore.webservice.helpers.DebugHelper: debug.threads by thread "diagnostics" (18):
"Reference Handler" daemon prio=10 Id=2 RUNNABLE
        at java.base@17.0.7/java.lang.ref.Reference.waitForReferencePendingList(Native Method)
        at java.base@17.0.7/java.lang.ref.Reference.processPendingReferences(Reference.java:253)
        at java.base@17.0.7/java.lang.ref.Reference$ReferenceHandler.run(Reference.java:215)
[...]

INFO  [2024-02-28 20:51:05,045] io.dockstore.webservice.helpers.DebugHelper: debug.database by thread "diagnostics" (18):
pool-size: 8
pool-active: 0
pool-idle: 8

INFO  [2024-02-28 20:51:05,046] io.dockstore.webservice.helpers.DebugHelper: debug.memory by thread "diagnostics" (18):
HEAP: init = 268435456(262144K) used = 70545688(68892K) committed = 144703488(141312K) max = 4294967296(4194304K)
NON-HEAP: init = 7667712(7488K) used = 121830168(118974K) committed = 123928576(121024K) max = -1(-1K)
POOL: CodeHeap 'non-nmethods', Non-heap memory
current: init = 2555904(2496K) used = 1457792(1423K) committed = 2555904(2496K) max = 5840896(5704K)
peak: init = 2555904(2496K) used = 1520384(1484K) committed = 2555904(2496K) max = 5840896(5704K)
collection: null
POOL: Metaspace, Non-heap memory
current: init = 0(0K) used = 84102304(82131K) committed = 84738048(82752K) max = -1(-1K)
peak: init = 0(0K) used = 84102304(82131K) committed = 84738048(82752K) max = -1(-1K)
collection: null
[...]
```
I've elided most of the thread dump, in total.   On my development box, it's around 10 lines per thread, and 34 threads are listed, totally approximately 25kB.  The memory pool dump is the next biggest entry, about 2.5kB total.

The `ps` output is not included here.  It's a ton of output on my development box, but is relatively small when running in ECS, a few lines.

My feeling is that, although it's intriguing to ponder logging this stuff to a different group, it might be more trouble than it's worth, if we can format the debug info so it doesn't overwhelm the other log info.

denis-yuen commented:
> * We could possibly modify our logging to condense each multi-line log entry to a single expandable entry: https://stackoverflow.com/questions/67396522/cloudwatch-multiline-log-messages-from-containerized-app-runnning-on-ecs-ec2  This would greatly reduce the visual impact of the logging entries that appear in this PR, as well as collapse the giant stack traces that currently appear in our logs.

I think this could work, in addition to swapping most of these outputs from `INFO` to `TRACE`
Presumably the other thing we can do is probably do some kind of filter to produce more than one logstream. 
The downside I think is we'll need to consider log expiry, but we probably need a ticket on that either way.

denis-yuen reviewed (commented):
cleaning list

svonworl commented on dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/DiagnosticsHelperTest.java:
```diff
@@ -0,0 +1,140 @@
+        confirmPassed("/This/Is/A/Test/Of/The/Censoring/System");
+        confirmPassed("a short decimal number: 7236463, a short hex number af081afb609d");
+        confirmCensored("f893b559e572f323b5369452752d56743970e39132136aff378b76492c7549dd");
+        confirmCensored("039832784664398210947231904282437420197091837413209843");
```
All of the key/credential/secret-like strings in this file are fake, I created them by banging on the keyboard and or encoding `/dev/random`.

svonworl commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CensorHelper.java:
```diff
@@ -0,0 +1,210 @@
+package io.dockstore.webservice.helpers;
```
I forgot to add copyright headers, will add them next time I push.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DiagnosticsHelper.java:
```diff
@@ -0,0 +1,387 @@
+        this.threadMXBean = ManagementFactory.getThreadMXBean();
+    }
+
+    public void start(Environment newEnvironment, SessionFactory newSessionFactory, DockstoreWebserviceConfiguration.DiagnosticsConfig newConfig) {
```
Just to be paranoid, is it worth making DiagnosticsHelper a singleton, and then ensuring this method can only be run once, like when you try to call `Thread.start()` twice on the same instance?

You don't have to implement this, and it is unlikely to happen the way our code is structured, but it did strike me when reviewing as the sort of thing you wouldn't want to accidentally instantiate multiple times.

svonworl replied:
In general, the singleton pattern can make the code harder to test and less flexible, so IMHO, we should only use it when it's critical that the code in question only be instantiated once.  This isn't one of those cases, if we launch DiagnosticsHelper twice, we'll see the double logging and fix it real quick, but nothing will get corrupted or otherwise messed up.

svonworl commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DiagnosticsHelper.java:
```diff
@@ -0,0 +1,387 @@
+        if (logger.isInfoEnabled()) {
+            Thread current = Thread.currentThread();
+            String message = String.format("diagnostics.%s by thread \"%s\" (%s):\n%s", type, current.getName(), current.getId(), valueSupplier.get());
+            logger.info(censorHelper.censor(message));
```
> Diagnostic information should probably be `trace`

That's fine if we can configure the logging system in our deployments to include this output, but not the other TRACE/DEBUG output.  The diagnostic output is most useful if it's always logged, so we can refer back to it to debug problems, but we don't want to cloud the logs with TRACE/DEBUG output from other modules.

denis-yuen replied:
> That's fine if we can configure the logging system in our deployments to include this output

This is pretty simple https://dev.to/cerveada/set-logback-to-different-level-for-specific-package-19db

> The diagnostic output is most useful if it's always logged 

This is configurable. If you have trace or debug on, these messages will always show up or vice versa. My intent here is to properly categorize the information 

i.e. `INFO` is normally pretty broad in scope `The INFO level designates informational messages that highlight the progress of the application at coarse-grained level.`
This sounds a lot more like `The TRACE Level designates finer-grained informational events than the DEBUG`
https://logging.apache.org/log4j/1.x/apidocs/org/apache/log4j/Level.html for example

denis-yuen replied:
Actually, `logback-test.xml` may be just when running tests locally. 
[dropwizard config ](https://www.dropwizard.io/en/release-3.0.x/manual/configuration.html#logging) has the same idea for customizing logging by package

svonworl replied:
> > That's fine if we can configure the logging system in our deployments to include this output
> 
> This is pretty simple https://dev.to/cerveada/set-logback-to-different-level-for-specific-package-19db
> 
> > The diagnostic output is most useful if it's always logged
> 
> This is configurable. If you have trace or debug on, these messages will always show up or vice versa. My intent here is to properly categorize the information
> 
> i.e. `INFO` is normally pretty broad in scope `The INFO level designates informational messages that highlight the progress of the application at coarse-grained level.` This sounds a lot more like `The TRACE Level designates finer-grained informational events than the DEBUG` https://logging.apache.org/log4j/1.x/apidocs/org/apache/log4j/Level.html for example

I think we differ philosophically here, I see `TRACE` and `DEBUG` as applying to tidbits of information that we're not interested in seeing most of the time, that we only enable to help us understand the behavior of some localized bit of code, whereas the information we're logging in this PR is relevant globally, will address some very significant blind spots, and help us to determine the cause of (and devise solutions to) a broad range of issues.  The additional information in this PR will help me _a lot_, with respect to both adding features and solving problems.

However, yes, we can change the log level to whatever the team wants, if we can make sure the display of the new information doesn't trigger the display of the less-important stuff...

denis-yuen replied:
Did the output change a lot from what you had documented in https://github.com/dockstore/dockstore/pull/5821#issuecomment-1969954529 ?

I'm seeing timing information, thread information, memory usage, periodic `ps` output and the helper is/was literally called "io.dockstore.webservice.helpers.DebugHelper" i.e. `debug`? 

>  if we can make sure the display of the new information doesn't trigger the display of the less-important stuff...

I also don't think it helps to think about it in terms of importance. Debugging info can be really important depending on what you're working on, so can `info`. In other contexts, they can both be noise.

Again, I'm more thinking about categorisation.  The convention is normally that stuff like heap vs non-heap memory, connection pool size, etc. is more about behind-the-scenes detail as this article puts it

> A INFO log is like the movie trailer, giving you the overall plot. DEBUG is the director's commentary, pointing out significant scenes and decisions. TRACE, on the other hand, includes every behind-the-scenes detail, including every script edit, camera angle, and production choice.

> Trace-level logs can reveal performance bottlenecks and optimization opportunities that might be invisible at higher logging levels:

> Monitoring Resource Utilization: Trace logs can show how resources like memory and CPU are used, helping you optimize them for better performance.

https://newrelic.com/blog/how-to-relic/why-you-need-trace-level-logging

svonworl replied:
Once again, I think there's philosophical differences at play.  I read your evidence and conclude "INFO", you read them and conclude "TRACE".  We're both smart people and sometimes we'll disagree, and that's ok.  There probably is no 100% "correct" answer, like tabs versus spaces, etc.

In the end, I'll set the level to whatever the team decides, and hope that we'll always include the new diagnostic output in the production logs, because a lot of the value of that information is that it's available after an unexpected failure of unknown origin.

svonworl replied:
So far, there's two "votes" for INFO level (me and the "approved" review) and one "vote" for TRACE level, so if we average them and round to nearest, we end up with DEBUG level.  So, I changed the `log` method to user the `Logger.debug` method.  Not complaining, but it's worth noting that it's a relatively costly change, because now, to merge, I'll need to update the logging config in compose_setup to ensure that the information shows up, do the compose_setup/dockstore-deploy PR dance again, and repeat testing on qa...

svonworl commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CensorHelper.java:
```diff
@@ -0,0 +1,210 @@
+ *
+ * 2. The number of "English points", where higher values represent sequences of characters more
+ *    characteristic of English language sequences, currently derived from letter-triplet
+ *    frequencies as calculated from an English text corpus.  Passages that resemble English text
```
> Trying to wrap my mind around this, how is this whole thing going to react in general to non-English secrets or secrets with accented characters?

The current implementation will censor longer base64-ish strings that look like keys/tokens (such as AWS keys and GitHub tokens), it's not designed to censor anything else.

denis-yuen replied:
But if I understand correctly, if this is using similarity to English to determine what should not be censored, wouldn't non-english text just look like random text and be more likely to be censored? 
i.e. log messages with non-english names, users with non-english names, etc.

svonworl replied:
The basic algorithm is something like:

1. Find "suspect" strings composed entirely of non-padding base64 (letter, numbers, `-`, `_`, `+`, and `/`) characters, of length 22 or more.
2. For each "suspect" string, calculate its score, and it's below a certain threshold, censor.

So, yeah, if there's a string that matches the criteria in Step 1 and doesn't look like English _at all_, it's likely to be censored.        Letter-frequency-wise, English is at least somewhat similar to many European languages, so candidate strings that are Spanish/French/German/etc will probably not be censored much of the time.  Strings that contain the higher portion of Latin-8 (common western characters with accents, etc) or the non-ASCII portion of Unicode (Asian characters, etc) will never be censored, because those aren't non-padding base64 characters, the strings won't be "suspects".

denis-yuen replied:
Oh, I see. So something needs to pass `BASE64_PATTERN` to be considered for censorship. 

May want to add
```
        confirmCensored("export BUNDLE_ENTERPRISE__CONTRIBSYS__COM=cafebabe:deadbeef");
        confirmCensored("export BUNDLE_ENTERPRISE__CONTRIBSYS__COM=secretPassword");
```
https://github.com/gitleaks/gitleaks which also has something a bit more off-the-shelf https://github.com/gitleaks/gitleaks/blob/master/config/gitleaks.toml

svonworl replied:
> Oh, I see. So something needs to pass `BASE64_PATTERN` to be considered for censorship.
> 
> May want to add
> 
> ```
>         confirmCensored("export BUNDLE_ENTERPRISE__CONTRIBSYS__COM=cafebabe:deadbeef");
>         confirmCensored("export BUNDLE_ENTERPRISE__CONTRIBSYS__COM=secretPassword");
> ```
> 
> https://github.com/gitleaks/gitleaks which also has something a bit more off-the-shelf https://github.com/gitleaks/gitleaks/blob/master/config/gitleaks.toml

`CensorHelper` won't censor either one of these strings, they're too regular and resemble English.

svonworl commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CensorHelper.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    private int characterClass(char c) {
+        if (c >= '0' && c <= '9') {
```
These comparisons are hardcoded here because it's super-critical that these methods do exactly what we want them to, otherwise the code will crash or malfunction.  We can use the classes/methods you suggest, but we'll be a little less sure that the delegated operations are what we want to happen.

denis-yuen replied:
Not sure I understand this, apache libraries are open source 
https://github.com/apache/commons-lang/blob/rel/commons-lang-3.14.0/src/main/java/org/apache/commons/lang3/CharSet.java#L72

svonworl commented on dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/DiagnosticsHelperTest.java:
```diff
@@ -0,0 +1,165 @@
+
+    @Test
+    void testCensoringWithRandomBase64Strings() {
+        Random random = new Random(0);
```
This test is deterministic, `Random `will generate the same stream of values given a particular seed, and the argument to the constructor (`0`) is the seed.

denis-yuen replied:
Neat. 
TIL `The legacy classes java.util.Random and java.util.SecureRandom were backfitted to implement this interface, so if you switch the declaration to the new interface, the legacy implementations become interchangeable with the new generators. The legacy classes’ algorithms were not improved, however, to avoid breaking Java’s ever-important backwards compatibility. In fact, given the same seed, the java.util.Random class in both Java 11 and Java 17 will output exactly the same sequence.`
https://blogs.oracle.com/javamagazine/post/java-pseudo-random-number-generator-enhancements

denis-yuen reviewed (changes requested):
Should probably return to discuss

david4096 reviewed (approved):
Awesome to see new diagnostic information in the logs! It would be nice-to-have that portions that are quantifiable had metrics defined in dockstore-deploy so we could track these over time/per container.

In my understanding the logs have the same access restrictions as the container and DB so censoring the logs doesn't seem critical. For maintainability I would consider not including that part, however it is good to have.

It might be preferable use an existing library (which I couldn't immediately find) or something like https://github.com/mazen160/secrets-patterns-db/blob/master/db/rules-stable.yml or https://github.com/gitleaks/gitleaks/blob/master/config/gitleaks.toml as linked by Denis.

denis-yuen reviewed (changes requested):
Hmmm, I think that puts a finger on two things that have been bugging me about this. I think I'll have to type up something longer in #dockstore-devs

svonworl commented:
Ok, I made a few changes, the current state is the proposed final version.

Part of my original "move it forward" plan was to remove the custom censoring code and replace it with the AWS Cloudwatch "data protection" feature.  However, upon deeper inspection, I learned that the AWS feature probably wouldn't catch the type of secrets that might leak, either because it's not looking for certain types of secrets (GitHub tokens, for example), or it requires they appear in certain contexts to detect them (for example, an AWS key needs to appear in proximity to the keyword `aws_secret_access_key` (or similar) for the AWS feature to censor it, which is unlikely to happen in our case).

By far, the most likely place for secrets to leak is the `ps` output.  So, to totally neutralize that scenario, I removed the `ps` output from this PR.  Now, the probability of a secret leaking is greatly reduced, much closer to the (low) background probability of leaking a secret in the rest of the logs, and censoring is much less important (either via our own code or the AWS feature).

So, I've removed the original webservice-side censoring code, and since the AWS data protection feature doesn't perform very well in our use case, I recommend that, at this juncture, we don't spend the time to enable it.

I should note that, left to my own devices, I would include the original webservice-side censoring code, but having dropped the `ps` output, I can merge this PR without it and not feel bad/worried/irresponsible/etc.

denis-yuen commented:
> ecause it's not looking for certain types of secrets (GitHub tokens, for example),

There is a custom data identifier feature https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-custom-data-identifiers.html
I think you should be able to add the github tokens since they start with gho, ghr, etc. https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/

denis-yuen commented:
> we don't spend the time to enable it.

I think it should only be a handful of steps https://aws.amazon.com/blogs/aws/protect-sensitive-data-with-amazon-cloudwatch-logs/ but you can create a follow-up ticket

denis-yuen reviewed (commented):
Would prefer to have at least the follow-up ticket for data protection
